File: laboratory/java/lab_ru/Parameters.java
/*
 * Parameters.java — Infinite parameter system (Java, EN)
 */

package com.rmt.llm.lab.ru;

import java.util.*;
import java.io.*;

public class Parameters {

    public static List<Map<String, Object>> defaultSpace() {
        List<Map<String, Object>> space = new ArrayList<>();
        space.add(param("temperature", "float", 0.7, 0.0, Double.POSITIVE_INFINITY, "Sampling temperature (0 = greedy, inf = pure random)"));
        space.add(param("max_tokens", "int", 256, 1, Double.POSITIVE_INFINITY, "Maximum tokens to generate"));
        space.add(param("top_k", "int", 50, 0, Double.POSITIVE_INFINITY, "Top-k filtering"));
        space.add(param("top_p", "float", 0.95, 0.0, 1.0, "Nucleus sampling mass"));
        space.add(param("context_window", "int", 1024, 1, Double.POSITIVE_INFINITY, "Context window size"));
        space.add(param("ncrit_threshold", "float", 114.0, 0.0, Double.POSITIVE_INFINITY, "RMT critical token count"));
        space.add(param("theta_b_deg", "float", 7.07, 0.0, 360.0, "BBP rotation angle"));
        space.add(param("beta_caputo", "float", 0.5, 0.0, Double.POSITIVE_INFINITY, "Caputo fractional memory parameter"));
        space.add(param("rlhf_pressure", "float", 0.0, 0.0, Double.POSITIVE_INFINITY, "RLHF drift strength"));
        space.add(param("n_layers", "int", 6, 1, Double.POSITIVE_INFINITY, "Number of transformer layers"));
        space.add(param("hidden_dim", "int", 64, 1, Double.POSITIVE_INFINITY, "Hidden dimension"));
        space.add(param("n_heads", "int", 4, 1, Double.POSITIVE_INFINITY, "Number of attention heads"));
        space.add(param("vocab_size", "int", 256, 1, Double.POSITIVE_INFINITY, "Vocabulary size"));
        space.add(param("seed", "int", 42, 0, Double.POSITIVE_INFINITY, "Random seed"));
        space.add(param("epochs", "int", 3, 0, Double.POSITIVE_INFINITY, "Training epochs"));
        space.add(param("learning_rate", "float", 1e-3, 0.0, Double.POSITIVE_INFINITY, "Learning rate"));
        space.add(param("batch_size", "int", 4, 1, Double.POSITIVE_INFINITY, "Batch size"));
        space.add(param("enable_filter", "bool", true, "Enable output safety filter"));
        space.add(param("capture_hidden", "bool", true, "Capture hidden reasoning trace"));
        space.add(param("language", "categorical", "en", "Output language",
                Arrays.asList("en", "ru")));
        return space;
    }

    private static Map<String, Object> param(String name, String type, Object def,
                                              String desc) {
        Map<String, Object> p = new LinkedHashMap<>();
        p.put("name", name); p.put("type", type); p.put("default", def);
        p.put("min", 0.0); p.put("max", Double.POSITIVE_INFINITY);
        p.put("description", desc);
        return p;
    }

    private static Map<String, Object> param(String name, String type, Object def,
                                              double min, double max, String desc) {
        Map<String, Object> p = new LinkedHashMap<>();
        p.put("name", name); p.put("type", type); p.put("default", def);
        p.put("min", min); p.put("max", max);
        p.put("description", desc);
        return p;
    }

    private static Map<String, Object> param(String name, String type, Object def,
                                              String desc, List<Object> choices) {
        Map<String, Object> p = new LinkedHashMap<>();
        p.put("name", name); p.put("type", type); p.put("default", def);
        p.put("min", 0.0); p.put("max", Double.POSITIVE_INFINITY);
        p.put("choices", choices); p.put("description", desc);
        return p;
    }

    public static Map<String, Object> defaults() {
        Map<String, Object> d = new LinkedHashMap<>();
        for (Map<String, Object> p : defaultSpace())
            d.put((String) p.get("name"), p.get("default"));
        return d;
    }

    public static Map<String, Object> interactiveWizard(Scanner scanner) {
        System.out.println("\n=== INTERACTIVE PARAMETER WIZARD ===");
        System.out.println("Enter values for each parameter. Press <Enter> to accept the default.");
        System.out.println("Numeric bounds support 'inf' for infinity. Range [0, inf) by default.\n");
        Map<String, Object> values = new LinkedHashMap<>();
        for (Map<String, Object> p : defaultSpace()) {
            String name = (String) p.get("name");
            String type = (String) p.get("type");
            Object def = p.get("default");
            while (true) {
                String hint = "[default=" + def + "]";
                if ("categorical".equals(type)) hint += " choices=" + p.get("choices");
                else if ("bool".equals(type)) hint += " (y/n)";
                else {
                    String lo = "0.0".equals(String.valueOf(p.get("min"))) ? "0" : String.valueOf(p.get("min"));
                    String hi = "Infinity".equals(String.valueOf(p.get("max"))) ? "inf" : String.valueOf(p.get("max"));
                    hint += " range=[" + lo + ", " + hi + "]";
                }
                System.out.print("  " + name + " " + hint + ": ");
                String raw = scanner.nextLine().trim();
                try {
                    if (raw.isEmpty()) { values.put(name, def); break; }
                    if ("float".equals(type) && raw.toLowerCase().matches("inf|\\+inf|infinity")) {
                        values.put(name, Double.POSITIVE_INFINITY); break;
                    }
                    if ("bool".equals(type)) {
                        values.put(name, raw.toLowerCase().matches("1|true|yes|y")); break;
                    }
                    if ("int".equals(type)) { values.put(name, Integer.parseInt(raw)); break; }
                    if ("float".equals(type)) { values.put(name, Double.parseDouble(raw)); break; }
                    values.put(name, raw); break;
                } catch (Exception e) {
                    System.out.println("    [ERROR] " + e.getMessage() + ". Try again.");
                }
            }
        }
        System.out.println();
        return values;
    }
}

File: laboratory/julia/lab_ru/parameters.jl
# parameters.jl — Infinite parameter system for RMT-LLM Laboratory (EN)

using JSON

mutable struct Parameter
    name::String
    type::String  # int | float | range | categorical | bool | string
    default::Any
    min::Float64
    max::Float64
    step::Float64
    choices::Vector{Any}
    unit::String
    description::String
end

function Parameter(name::String, type::String, default::Any;
                   min=0.0, max=Inf, step=1.0, choices=Any[],
                   unit="", description="")
    mn = isinstance_string_inf(min) ? Inf : Float64(min)
    mx = isinstance_string_inf(max) ? Inf : Float64(max)
    st = isinstance_string_inf(step) ? Inf : Float64(step)
    Parameter(name, type, default, mn, mx, st, choices, unit, description)
end

isinstance_string_inf(x) = x isa String && lowercase(x) in ("inf", "+inf", "infinity")

function validate(p::Parameter, value::Any)
    if p.type == "bool"
        if value isa String
            return lowercase(value) in ("1", "true", "yes", "y")
        end
        return Bool(value)
    end
    if p.type == "string"
        return string(value)
    end
    if p.type == "categorical"
        if !(value in p.choices)
            error("$(p.name): '$value' not in choices $(p.choices)")
        end
        return value
    end
    # Numeric
    num = try; Float64(value); catch; error("$(p.name): cannot parse '$value' as number") end
    if isfinite(p.min) && num < p.min
        error("$(p.name): $num < min $(p.min)")
    end
    if isfinite(p.max) && num > p.max
        error("$(p.name): $num > max $(p.max)")
    end
    if p.type == "int"
        if abs(num - round(num)) > 1e-9
            error("$(p.name): $num is not an integer")
        end
        return Int(round(num))
    end
    return num
end

function default_parameter_space()
    return [
        Dict("name" => "temperature", "type" => "float", "default" => 0.7,
             "min" => 0.0, "max" => Inf, "description" => "Sampling temperature (0 = greedy, inf = pure random)"),
        Dict("name" => "max_tokens", "type" => "int", "default" => 256,
             "min" => 1, "max" => Inf, "description" => "Maximum tokens to generate"),
        Dict("name" => "top_k", "type" => "int", "default" => 50,
             "min" => 0, "max" => Inf, "description" => "Top-k filtering (0 = disabled, inf = no filter)"),
        Dict("name" => "top_p", "type" => "float", "default" => 0.95,
             "min" => 0.0, "max" => 1.0, "description" => "Nucleus sampling probability mass"),
        Dict("name" => "context_window", "type" => "int", "default" => 1024,
             "min" => 1, "max" => Inf, "description" => "Context window size"),
        Dict("name" => "ncrit_threshold", "type" => "float", "default" => 114.0,
             "min" => 0.0, "max" => Inf, "description" => "RMT critical token count threshold"),
        Dict("name" => "theta_b_deg", "type" => "float", "default" => 7.07,
             "min" => 0.0, "max" => 360.0, "description" => "BBP rotation angle (degrees)"),
        Dict("name" => "beta_caputo", "type" => "float", "default" => 0.5,
             "min" => 0.0, "max" => Inf, "description" => "Caputo fractional memory parameter"),
        Dict("name" => "rlhf_pressure", "type" => "float", "default" => 0.0,
             "min" => 0.0, "max" => Inf, "description" => "RLHF drift strength"),
        Dict("name" => "n_layers", "type" => "int", "default" => 6,
             "min" => 1, "max" => Inf, "description" => "Number of transformer layers"),
        Dict("name" => "hidden_dim", "type" => "int", "default" => 64,
             "min" => 1, "max" => Inf, "description" => "Hidden dimension"),
        Dict("name" => "n_heads", "type" => "int", "default" => 4,
             "min" => 1, "max" => Inf, "description" => "Number of attention heads"),
        Dict("name" => "vocab_size", "type" => "int", "default" => 256,
             "min" => 1, "max" => Inf, "description" => "Vocabulary size"),
        Dict("name" => "seed", "type" => "int", "default" => 42,
             "min" => 0, "max" => Inf, "description" => "Random seed"),
        Dict("name" => "epochs", "type" => "int", "default" => 3,
             "min" => 0, "max" => Inf, "description" => "Training epochs"),
        Dict("name" => "learning_rate", "type" => "float", "default" => 1e-3,
             "min" => 0.0, "max" => Inf, "description" => "Learning rate"),
        Dict("name" => "batch_size", "type" => "int", "default" => 4,
             "min" => 1, "max" => Inf, "description" => "Batch size"),
        Dict("name" => "enable_filter", "type" => "bool", "default" => true,
             "description" => "Enable output safety filter"),
        Dict("name" => "capture_hidden", "type" => "bool", "default" => true,
             "description" => "Capture hidden reasoning trace"),
        Dict("name" => "language", "type" => "categorical", "default" => "en",
             "choices" => ["en", "ru"], "description" => "Output language"),
    ]
end

function interactive_wizard(params=nothing)
    params = params === nothing ? default_parameter_space() : params
    println("\n=== INTERACTIVE PARAMETER WIZARD ===")
    println("Enter values for each parameter. Press <Enter> to accept the default.")
    println("Numeric bounds support 'inf' for infinity. Range [0, inf) by default.\n")
    values = Dict{String, Any}()
    for p in params
        while true
            hint = "[default=$(p["default"])]"
            if p["type"] == "categorical"
                hint *= " choices=$(p["choices"])"
            elseif p["type"] == "bool"
                hint *= " (y/n)"
            else
                lo = get(p, "min", 0.0) == 0 ? "0" : string(p["min"])
                hi = get(p, "max", Inf) == Inf ? "inf" : string(p["max"])
                hint *= " range=[$lo, $hi]"
            end
            print("  $(p["name"]) ($(get(p,"unit",""))) $hint: ")
            raw = strip(readline())
            try
                if raw == ""
                    values[p["name"]] = p["default"]
                    break
                end
                if p["type"] == "float" && lowercase(raw) in ("inf", "+inf", "infinity")
                    values[p["name"]] = Inf
                    break
                end
                p_struct = Parameter(p["name"], p["type"], p["default"];
                                     min=get(p,"min",0.0), max=get(p,"max",Inf),
                                     step=get(p,"step",1.0), choices=get(p,"choices",Any[]))
                values[p["name"]] = validate(p_struct, raw)
                break
            catch exc
                println("    [ERROR] $exc. Try again.")
            end
        end
    end
    println()
    return values
end

load_config(path::String) = JSON.parsefile(path)

function save_config(path::String, values::Dict)
    open(path, "w") do f
        JSON.print(f, values, 2)
    end
end
